### ISSUE TITLE
attention diffusion

### ISSUE THREAD

**Issue by falseywinchnet** [karpathy/nanoGPT#595]
for a variety of reasons, attention is the best place to get it on.
```
class CausalSelfAttention(nn.Module):

    def __init__(self, config):
        super().__init__()
        assert config.n_embd % config.n_head == 0
        # key, query, value projections for all heads, but in a batch
        self.c_attn = nn.Linear(config.n_embd, 3 * config.n_embd, bias=config.bias)
        # output projection
        self.c_proj = nn.Linear(config.n_embd, config.n_embd, bias=config.bias)
        # regularization
        self.attn_dropout = nn.Dropout(config.dropout)
        self.resid_dropout = nn.Dropout(config.dropout)
        self.n_head = config.n_head
        self.n_embd = config.n_embd
        self.dropout = config.dropout
        # flash attention makes GPU go brrrrr but support is only in PyTorch >= 2.0
        self.flash = hasattr(torch.nn.functional, 'scaled_dot_product_attention')
        if not self.flash:
            print("WARNING: using slow attention. Flash Attention requires PyTorch >= 2.0")
            # causal mask to ensure that attention is only applied to the left in the input sequence
            self.register_buffer("bias", torch.tril(torch.ones(config.block_size, config.block_size))
                                        .view(1, 1, config.block_size, config.block_size))

    def forward(self, x):
        B, T, C = x.size()  

        # Compute Q, K, V for all heads in batch
        q, k, v = self.c_attn(x).split(self.n_embd, dim=2)

        # Reshape into multi-head format
        q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2)  
        k = k.view(B, T, self.n_head, C // self.n_head).transpose(1, 2)  
        v = v.view(B, T, self.n_head, C // self.n_head).transpose(1, 2)  

        # Generate Gaussian noise per head
        noise = torch.randn_like(q) * 0.1  # Small variance perturbation

        # Two perturbations: (+η) and (-η)
        q_pos, k_pos, v_pos = q + noise, k + noise, v + noise
        q_neg, k_neg, v_neg = q - noise, k - noise, v - noise

        # Compute attention twice
        if self.flash:
            att_pos = torch.nn.functional.scaled_dot_product_attention(q_pos, k_pos, v_pos, attn_mask=None, dropout_p=self.dropout if self.training else 0, is_causal=True)
            att_neg = torch.nn.functional.scaled_dot_product_attention(q_neg, k_neg, v_neg, attn_mask=None, dropout_p=self.dropout if self.training else 0, is_causal=True)
        else:
            att_pos = (q_pos @ k_pos.transpose(-2, -1)) * (1.0 / math.sqrt(k_pos.size(-1)))
            att_pos = att_pos.masked_fill(self.bias[:, :, :T, :T] == 0, float('-inf'))
            att_pos = F.softmax(att_pos, dim=-1)
            att_pos = self.attn_dropout(att_pos)
            att_pos = att_pos @ v_pos

            att_neg = (q_neg @ k_neg.transpose(-2, -1)) * (1.0 / math.sqrt(k_neg.size(-1)))
            att_neg = att_neg.masked_fill(self.bias[:, :, :T, :T] == 0, float('-inf'))
            att_neg = F.softmax(att_neg, dim=-1)
            att_neg = self.attn_dropout(att_neg)
            att_neg = att_neg @ v_neg

        # Average both attention outputs to cancel noise influence idea stolen from a paper on ITD and windmills
        y = (att_pos + att_neg) / 2  

        # Reshape back to standard format
        y = y.transpose(1, 2).contiguous().view(B, T, C)

        # Output projection
        y = self.resid_dropout(self.c_proj(y))
        return y

```

- attention is the only truly numeric place in the model, the rest has some semantic stuff
- balanced doubling means its invisible to backprop or might be, its fast anyway
- may help models generalize and learn
- something something optimal transport theory 

thoughts?

@kyunghyuncho @karpathy

**Comment by falseywinchnet** [comment]
I propose two further improvements that greatly improve learning, in my repo.



 
​

**Comment by falseywinchnet** [comment]
A Model 1 that learns dropout/mask fill.
A LatentRNN that expands embeddings in latent space.
A Model 2 that selects/refines final outputs and produces a distribution over tokens.

Here's the idea:
model 1 is a many head, few layer model with more emb dim and a shorter seq length, say, 64. Multiple such models can be used and their output overlapped- we'll get to that in a moment.
it produces a latent representation and doesn't select with an MLP.
its trained with dropout/masking against source material to refine it.
some of its predictions are dropped. they are replaced with top-k predictions made by an RNN operating in latent space, which is trained on complete predictions made by model 1. in other words, model 1 is allowed to embed content without masking, and the rnn is trained to match the embeddings.

the RNN predictions appended to the model 1 partial output are fed
to a second deeper layer, fewer head model with a longer seq len.

that model then selects most likely predictions and these are unembedded and trained into model 1 as if inputs.
the second model is back-propped on loss for final predictions in the normal manner.

during evaluation, no masking is applied, and the char-rnn provides continuation of sequence produced by first model.


recurrent reasoning: a latent "informational sequence" is injected between the model1 output and RNN's prediction in latent space and then removed from what Is sent to model 2. model 2's predictions are inserted into this space, essentially, appended to what the RNN sees. the RNN then changes its output because of this, and then model 2's predictions change. Therefore, the second model predicts a longer sequence using a reoccurring reasoning step where on each iteration it feeds into the buffer one character as top-k, and char-rnn continues from it, and the model2 then thusly learns how to discipline the char-rnn outputs.

the first model proposes concepts, the rnn hallucinates conclusions, the third model realizes clarification, the first model trains the second model, the third model and first model are trained by the data, the first model is trained by the third model, and the third model is trained by the second.

**Comment by falseywinchnet** [comment]
one last topic for the day to tie it all in together. we could add phase embedding information in the beginning. it then introduces a continuity dimension or relational component that model2 learns to anticipate being present. in other words, here, its an auxiliary loss because otherwise disciplining it is costly. the rnn is trained to predict it alongside a prediction of the underlying tok+pos embeddings. in other words, the rnn must learn to generate phase information- and cannot do so- but the second model can learn to critique based on it- and expects it to be present. the inadvertent lying parrot situation is reconciled by consensus information. signals are present to communicate truthiness. what has to happen, I reckon, is that the first models are in fact a many-head transformer block, and the x2 is processed independently. these then transform the embeddings respectively. the rnn, conversely, must take and produce complex results, and the complex results are then decomposed into real and imaginary embeddings as x and x2 and fed into the second stage model.

Reasoned for a few seconds
Below is a final conceptual “tying it all together” that integrates your ideas about phase embeddings, consensus signals, and the multi-model pipeline. It shows how you can introduce a complex or “dual stream” representation (real + imaginary) and have each component (Model 1, RNN, Model 2) handle it in a coherent way—ultimately to reduce “lying parrot” outputs and encourage internal consistency.

1. Phase Embeddings: A Quick Refresher

The phase embedding concept treats part of the embedding dimension as a sort of “imaginary axis.”
You can think of each token embedding (or the entire sequence embedding) as a complex number 
(
real
,
imag
)
(real,imag).
This “phase” dimension encodes relational or continuity information, e.g., how the position of one token relates to others, or how an entire chunk lines up contextually.
Why Phase Embeddings Help
Continuity Dimension
It introduces an extra “relational channel” that can encode whether new content is consistent with existing context.
Consensus Checking
A downstream model (Model 2) can critique whether the real–imag parts are coherent.
If the RNN (or earlier model) tries to produce “nonsense imaginary parts,” the next model can detect mismatch between real and imaginary streams.
2. Incorporating Phase Embeddings in the Multi-Model Pipeline

Let’s see how each step (Model 1, RNN, Model 2) deals with real/imag embeddings:

2.1 Model 1 (Two-Stream Transformer Block)

Many heads, few layers, short chunk – as before. But it now:
Maintains real embeddings (
x
x) and phase embeddings (
x
2
x2) for each token.
Possibly uses ResBlock or CausalSelfAttentionPair that processes 
(
x
,
x
2
)
(x,x2) in parallel streams.
Is trained with heavy dropout/masking on both real and imaginary components.
Output of Model 1
A short sequence of embeddings: 
(B, T, C)
(B, T, C) for real + 
(B, T, C)
(B, T, C) for imaginary.
Or you can treat them as a single 
(B, T, 2C)
(B, T, 2C) or “complex” representation.
Let’s call them 
(
x
real
,
x
imag
)
(x 
real
​	
 ,x 
imag
​	
 ).
2.2 Latent RNN (Hallucination of Both Real + Imag)

The RNN sees the combined real + imaginary streams (or a single “2C”-dim vector).
Its job: predict the next embeddings in both streams.
It might do so step-by-step (like an autoregressive approach) or chunk-wise.
The important twist: it cannot cheat by ignoring the imaginary dimension, because:
It is trained on real data from Model 1’s unmasked outputs, including both real and imag parts.
Model 2 expects coherent real–imag pairs. If the RNN produces random imaginary vectors that break continuity, Model 2 can penalize or ignore those expansions.
Reducing “Lying Parrot” Behaviors
Because the RNN must produce consistent real + imag streams that match Model 1’s style, it’s less likely to generate “nonsense” expansions—any mismatch stands out in the phase dimension.
2.3 Model 2 (Deeper, Fewer Heads, Large Seq)

Receives the combined real + imaginary expansions from the RNN (plus the short chunk from Model 1).
Critiques / Selects expansions by checking:
Are the real and imag components aligned?
Do they match known patterns (tokens, positions, continuity)?
Is the next-token prediction consistent overall?
Then Model 2 produces the final tokens (or a final embedding → unembedding → tokens).
Trained on standard next-token cross-entropy, but crucially, it can see the “phase channel” and expect it to be coherent with the real channel.
This effectively becomes an auxiliary consistency check: if the imaginary part doesn’t match the real part’s location, or if the new token is out of sync, Model 2’s output distribution shifts.
3. Handling “Consensus Information”

You mentioned “signals are present to communicate truthiness” – i.e., the system encodes meta-information about whether an embedding is consistent or not. The phase dimension can serve that role:

Model 1:
Embeds local chunk tokens into 
(
x
,
x
2
)
(x,x2). The phase part can reflect relative positions, continuity signals, or any “semantic alignment” dimension.
RNN:
Has to continue that phase consistently. If it tries to produce contradictory imaginary embeddings, the next stage can detect a mismatch.
Model 2:
Learns to interpret and penalize those mismatches, effectively imposing a “truthiness” check on the expansions.
4. Iterative Reasoning, Real + Imag Streams

You also mention “the second stage model’s predictions are appended to the space… the rnn changes its output… the second model changes.” In complex (or real+imag) form:

Iteration:
Model 2 picks a token (or a small chunk).
That’s unembedded into 
(
Δ
x
,
Δ
x
2
)
(Δx,Δx2).
The RNN sees this appended chunk in the real–imag domain, updates its “hallucination” forward.
Model 2 reevaluates whether the new RNN expansions remain coherent.
Repeat.
Why This Reduces Lying
Each iteration’s partial expansion must keep real and imag streams coherent enough for Model 2’s final acceptance.
If the imaginary part says “token belongs to position i,” but the real part claims a different adjacency, it’s flagged.


this secondary phase embedding is produced as follows:

def compute_phase_embedding(self, tok_emb, pos_emb):
            """
            Computes a secondary embedding for each token by accumulating phase shifts.
            
            For each token position i in x (of shape (B, T, C)):
              - We have two positional embeddings:
                   * A forward rope embedding for tokens j > i.
                   * A backward rope embedding for tokens j < i.
              - For each token j (j != i), we compute a positional indication:
                   If j > i: use forward embedding (function of j - i)
                   If j < i: use backward embedding (function of i - j)
              - Multiply that positional indication (elementwise) by x[j] to get a phase shift contribution.
              - Sum over all j to obtain a cumulative phase shift, φ[i].
              - Form a complex number with:
                   real part = x[i]
                   imaginary part = φ[i]
                and take the magnitude:
                   x2[i] = sqrt( x[i]^2 + φ[i]^2 )
            
            Args:
                tok_emb: Tensor of shape (B, T, C) representing token embeddings.
                pos_emb: Tensor of shape (T,C) with positions (0, 1, ..., T-1).
            
            Returns:
                x2: Tensor of shape (B, T, C), the secondary embeddings.
            """
            B, T, C = tok_emb.shape
            device = tok_emb.device
        
            # Define fixed frequencies per dimension (like standard RoPE frequencies).
            dims = torch.arange(C, device=device, dtype=torch.float32)  # (C,)
            freqs = 1.0 / (10000 ** (dims / C))  # (C,)
        
            # Compute a matrix of relative offsets: offset[i, j] = j - i.
            indices = torch.arange(T, device=device)
            offset_matrix = indices.unsqueeze(0) - indices.unsqueeze(1)  # shape (T, T)
        
            # Create masks for forward and backward directions.
            forward_mask = (offset_matrix > 0).float()  # (T, T)
            backward_mask = (offset_matrix < 0).float()  # (T, T)
        
            # For each pair (i, j) compute the absolute offset:
            # For forward, we care about (j - i) and for backward, (i - j).
            forward_offsets = torch.clamp(offset_matrix, min=0)  # (T, T): zeros for j <= i.
            backward_offsets = torch.clamp(-offset_matrix, min=0)  # (T, T): zeros for j >= i.
        
            # Compute forward and backward positional rope embeddings.
            # Here we use a simple sinusoidal function.
            # Each will have shape (T, T, C): for each i and j, a vector of length C.
            forward_pos_emb = torch.sin(forward_offsets.unsqueeze(-1) * freqs.view(1, 1, C))
            backward_pos_emb = torch.sin(backward_offsets.unsqueeze(-1) * freqs.view(1, 1, C))
            
            # Combine them: for each pair (i,j), use the forward embedding if j>i, backward if j<i.
            phase_emb = forward_mask.unsqueeze(-1) * forward_pos_emb + backward_mask.unsqueeze(-1) * backward_pos_emb
            # For j == i, offset is 0 so sin(0)=0; effectively these contribute nothing.
        
            # Now, for each token position i (for each batch element), accumulate contributions from all j:
            # φ[i, c] = sum_{j != i} [ phase_emb[i,j,c] * x[b,j,c] ]
            # Using einsum: treat phase_emb as (T, T, C) and x as (B, T, C) so that we sum over j.
            phi = torch.einsum('ijc,bjc->bic', phase_emb, tok_emb)
            phi = phi / (T ** 0.5)
            return phi

it is then used however you see fit, however, I recommend reoccurently diluting it by adding a normalized copy of x to it and reusing it in multiple stages. alone it's too powerful and overpowers the localized signal. but that's in simpler models like straightforward gpt.



these principles may offer us a more lightweight form of recursion than, for example, "Scaling up Test-Time Compute with Latent Reasoning:
A Recurrent Depth Approach" @JonasGeiping

they may also take us closer to a productive theory of mind approach where the model behaves more like a human mind, internally presenting, predicting, selecting, and refining, and learning at all stages to better do what we intend for it.

**Comment by falseywinchnet** [comment]
Proposed Architecture Overview
We propose extending the standard transformer-based language model by incorporating a Variational Recurrent Neural Network (VRNN) that generates a branching tree of possible continuations at each step, constrained by top-k sampling. Specifically, given a sequence of length N and a top-k parameter k, the model enumerates up to k^N potential forward paths. In practice, if k equals 5 and the sequence length under consideration is 5, the resulting structure contains 5^5 expansions. The hidden states, along with these top-k predictions, are stored in a multidimensional representation (for instance, shaped approximately by (B, k^N, ...)), where B denotes the batch dimension and additional dimensions capture the relevant features and internal states.

Stage 1: VRNN-Enhanced Generation

Top-k Embeddings: At each time step i, a “first-stage” submodule identifies the top-k candidate embeddings for the next token.
VRNN Prediction Tree: These k embeddings serve as inputs to the VRNN, which then generates a branching tree of plausible continuations for subsequent tokens.
Prepended Predictions: The predictions from the first-stage submodule are prepended to each branching path. This results in a representation of both immediate and longer-range predictions.
Stage 2: Attention-Based Re-weighting
Once the VRNN has proposed multiple forward paths, a transformer-based attention mechanism processes the prior context, the current candidate prediction, and each candidate forward sequence. The goal is to refine the likelihood scores associated with each predicted token at the current step, effectively steering the model toward more coherent continuations. Attention mechanisms re-weight each path by incorporating long-range dependencies and the contextual fit of each branching sequence.

**Comment by iabbas09** [comment]
Hardcoded noise scale 0.1 → should be a config param.

Comments like "unclear what alpha to use" and "idea stolen from a paper" need to be cleaned up.

Add a short comment or doc link to support the theoretical motivation (variance regularization / optimal transport).

Suggest gating with if config.attn_noise: to make this optional.

Let me know if you want help cleaning it up for a PR.
